src/lfview/lf-data-provider.ts

```diff
@@ -0,0 +1,487 @@
+            const projectUri = splittedUri.slice(0, - this.path_offset + 3).join('/') + '/';
+            const library_root = new LFDataProviderNode(projectLabel, projectUri, LFDataProviderNodeRole.ROOT, this.type, []);
+            root.children.push(library_root);
+            return library_root;
+        }
+        return existingLibraryRoot;
+    }
+
+    /**
+     * Opens the file associated with the node in the Tree View.
+     * @param node - The node whose file is to be opened.
+     * @param beside - Whether to open the file beside the current editor.
+     */
+    goToFileCommand(node: LFDataProviderNode, beside: boolean) {
+        vscode.workspace.openTextDocument(node.uri).then(doc => {
+            vscode.window.showTextDocument(doc, beside == false ? undefined : vscode.ViewColumn.Beside).then(e => {
+                if (node.position) {
+                    const selection = this.getHighlightSelection(node);
+                    e.selection = selection;
+                    e.revealRange(new vscode.Range(
+                        selection.anchor.line,
+                        selection.anchor.character,
+                        selection.active.line,
+                        selection.active.character),
+                        vscode.TextEditorRevealType.InCenter);
+                }
+            });
+        });
+    }
+
+    /**
+     * Imports the reactor associated with the selected LFDataProviderNode.
+     * @param node - The node representing the reactor to import.
+     */
+    async importReactorCommand(node: LFDataProviderNode) {
+        const editor = vscode.window.activeTextEditor;
+        if (editor) {
+            if (!editor.document.uri.fsPath.endsWith('.lf')) {
+                vscode.window.showErrorMessage('The active editor must be a Ligua Franca program.');
+                return;
+            }
+            const relativePath = this.getRelativePath(editor.document.uri.fsPath, node.uri.fsPath);
+            const importText = `import ${node.label.toString()} from "${relativePath}"\n`;
+            const position = await this.getTargetPosition(editor.document.uri);
+            this.addTextOnActiveEditor(editor, position.end, importText);
+        }
+    }
+
+    /**
+     * Gets the selection to highlight in the editor for the node.
+     * @param node - The node for which to get the selection.
+     * @returns The selection to highlight.
+     */
+    getHighlightSelection(node: LFDataProviderNode): vscode.Selection {
+        const editor = vscode.window.activeTextEditor;
+        if (editor) {
+            const sel = new vscode.Selection(node.position.start - 1, 0, node.position.start - 1, this.HIGHLIGHT_OFFSET);
+            const selectionRange = new vscode.Range(sel.start.line, sel.start.character, sel.end.line, sel.end.character);
+            const highlighted = editor.document.getText(selectionRange);
+            const idx = highlighted.indexOf(node.label.toString());
```

Comment:
What is `node.label`? Searching for a string in order to determine where the selection should end seems like a code smell to me. Similar uses of ad-hoc string matching have bitten us in the past.

Reply:
is actually the name of the reactor returned by the Language Server. My idea here is to check the index of the reactor’s name to highlight it. I’m not a fan of ad-hoc string matching either, but in this case, I thought we wouldn’t have ambiguity problems for two reasons:
1. There cannot be two reactors with the same name.
2. I look for the string in a specific selection, which is the one I get using the position of the node.

However, I’m completely open to your suggestions.

Reply:
Well, I guess the "right" way to do this would probably be to query the CST using `NodeModelUtils`, which involves a call into the language server. The selection could also contain comments that have the same text as the reactor name, for example, or the reactor name might not be within `HIGHLIGHT_OFFSET` of the start position. Another option would be to highlight the entire node instead of just the reactor name, so that at least the behavior does not have these weird corner cases. However, I'm also OK with leaving this as-is in order to save time since the bad behaviors that are possible are not too bad.

Reply:
Initially, I implemented this function by highlighting the entire reactor. However, I noticed that selecting the reactor on the diagram only highlights its name in the code. I wanted to replicate this behavior but couldn’t find how the diagrams achieve it, so I attempted my own solution. If you know where this implementation is and can guide me, that would be helpful. Otherwise, I think I’ll revert to highlighting the entire reactor, which seems to be a more robust solution

Reply:
@NiklasRentzCAU can you leave a comment here that explains how "diagram selects text" works and were and how it is implemented?

@vinzbarbuto I remember from a meeting that Edward disliked that the diagram did not select the whole text in VS Code while this is the case in Eclipse. @NiklasRentzCAU is there a straightforward way to make this configurable in klighd-vscode/Klighd?

Reply:
> @vinzbarbuto I remember from a meeting that Edward disliked that the diagram did not select the whole text in VS Code while this is the case in Eclipse. 

Yes, I believe the best solution is to highlight the entire reactor text. This approach avoids edge cases and is quite robust.

Reply:
What is highlighted is implemented by Sprotty on the Java-Server-Side in the [DiagramSelectionListener](https://github.com/eclipse-sprotty/sprotty-server/blob/0ab2731bf9e3152a53d7ab1cd18e3fa832d9fd26/org.eclipse.sprotty.xtext/src/main/java/org/eclipse/sprotty/xtext/DiagramSelectionListener.xtend#L58). Here you can see that Sprotty defaults to using the "significant region" to highlight in most cases only the name of the state/region/reactor/whatever instead of the entire text. If there is no such "significant region" defined in the Xtext grammar, it will fall back to using the entire text instead, see Sprotty's [TextRegionProvider](https://github.com/eclipse-sprotty/sprotty-server/blob/0ab2731bf9e3152a53d7ab1cd18e3fa832d9fd26/org.eclipse.sprotty.xtext/src/main/java/org/eclipse/sprotty/xtext/tracing/TextRegionProvider.xtend#L48).

In KLighD we slightly extended this default behavior to only generate the traces lazily, as that would otherwise increase the diagram generation time. We do that as Sprotty intended by extending the class (see [here](https://github.com/kieler/KLighD/blob/86f5b925a426298b0f7b0ae5ccd892e82b0b4a17/plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/utils/KeithDiagramSelectionListener.xtend#L50-L77)) and putting it into our used module for the diagrams (see [here](https://github.com/kieler/KLighD/blob/86f5b925a426298b0f7b0ae5ccd892e82b0b4a17/plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphDiagramModule.xtend#L72-L74)). This behavior can be changed by supplying a different module to the language server like we do it [here](https://github.com/kieler/KLighD/blob/86f5b925a426298b0f7b0ae5ccd892e82b0b4a17/plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/launch/AbstractLsCreator.xtend#L79-L100) by default and which can be changed for LF in [this file](https://github.com/lf-lang/lingua-franca/blob/790fb5b94551a94f952c5cc9345f11f6cb653f36/lsp/src/main/java/org/lflang/diagram/lsp/LanguageDiagramServer.java#L36-L52).
(Btw., the fix that you currently use there is not necessary anymore, that has been released in an older KLighD version that you do already use.)